Sources/NIOFileSystem/FileSystem.swift

```diff
@@ -0,0 +1,1442 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Atomics
+import NIOCore
+@preconcurrency import SystemPackage
+
+#if canImport(Darwin)
+import Darwin
+#elseif canImport(Glibc)
+import Glibc
+#endif
+
+/// A file system which interacts with the local system. The file system uses a thread pool to
+/// perform system calls.
+///
+/// ### Creating a `FileSystem`
+///
+/// You should prefer using the ``FileSystem/shared`` instance of the file system. The
+/// ``FileSystem/shared`` instance uses two threads unless the `SWIFT_FILE_SYSTEM_THREAD_COUNT`
+/// environment variable is set.
+///
+/// If you require more granular control you can create a ``FileSystem`` with the required number
+/// of threads by calling ``withFileSystem(numberOfThreads:_:)``.
+///
+/// ### Errors
+///
+/// Errors thrown by ``FileSystem`` will be of type:
+/// - ``FileSystemError`` if it wasn't possible to complete the operation, or
+/// - `CancellationError` if the `Task` was cancelled.
+///
+/// ``FileSystemError`` implements `CustomStringConvertible`. The output from the `description`
+/// contains basic information including the error code, message and underlying error. You can get
+/// more information about the error by calling ``FileSystemError/detailedDescription()`` which
+/// returns a structured multi-line string containing information about the error.
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+public final class FileSystem: Sendable, FileSystemProtocol {
```

Comment:
Does this actually need to be a `class`?

Reply:
@glbrntt you never replied here

Reply:
Oh sorry, I missed this one. Just changed it.